lib/basedriver/capabilities.js

```diff
@@ -0,0 +1,111 @@
+import _ from 'lodash';
+import { validator } from './desired-caps';
+import { util } from 'appium-support';
+
+// Takes primary caps object and merges it into a secondary caps object.
+// (see https://www.w3.org/TR/webdriver/#dfn-merging-capabilities)
+function mergeCaps (primary = {}, secondary = {}) {
+  let result = Object.assign(primary);
+
+  for (let [name, value] of _.toPairs(secondary)) {
+    // Overwriting is not allowed. Primary and secondary must have different properties (w3 rule 4.4)
+    if (!_.isUndefined(primary[name])) {
+      throw new Error(`property ${name} should not exist on both primary and secondary`);
+    }
+    result[name] = value;
+  }
+
+  return result;
+}
+
+// Validates caps against a set of constraints
+function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
+  let message = '';
+  if (!_.isObject(caps)) {
+    throw new Error(`must be a JSON object`);
```

Comment:
This error gets caught by other functions and then the error message is concatenated so it will look like `"Invalid argument 'capabilities': must be a JSON object"`